### PR TITLE
Document Traefik setup for Syncthing discosrv

### PR DESCRIPTION
**Pull request by Cobertos**
Traefik's setup was convoluted, especially for rewriting headers to get `X-Client-Port`. That's why I broke it up in the docs by header

### CODE REVIEW

File: users/stdiscosrv.rst [hunk 1]
```diff
@@ -346,6 +346,99 @@ page. Note that that page is directed at setting up a proxy for the
 Syncthing web UI. You should do the proper path and port adjustments to proxying
 the discovery server and your particular setup.
 
+Traefik 2
+"""""""""
+
+Traefik will send ``X-Forwarded-For`` by default
```
Comment (tomasz1986) [inline users/stdiscosrv.rst:352]:
```suggestion
Traefik will send ``X-Forwarded-For`` by default.
```

File: users/stdiscosrv.rst [hunk 2]
```diff
@@ -346,6 +346,99 @@ page. Note that that page is directed at setting up a proxy for the
...
+
+To send ``X-Forwarded-Tls-Client-Cert``
```
Comment (tomasz1986) [inline users/stdiscosrv.rst:354]:
```suggestion
To send ``X-Forwarded-Tls-Client-Cert``:
```

File: users/stdiscosrv.rst [hunk 3]
```diff
@@ -346,6 +346,99 @@ page. Note that that page is directed at setting up a proxy for the
...
+        labels:
+          - "traefik.enable=true"
+          - "traefik.http.middlewares.syncthing-discosrv-middleware.passtlsclientcert.pem=true"
+          - "traefik.http.services.syncthing-discosrv.loadbalancer.server.port=8443"
+          - "traefik.http.routers.syncthing-discosrv.entrypoints=https"
+          - "traefik.http.routers.syncthing-discosrv.rule=Host(`st-ds.xxx.dev`)"
+          - "traefik.http.routers.syncthing-discosrv.tls.options=syncthing-discosrv@file"
+          - "traefik.http.routers.syncthing-discosrv.middlewares=syncthing-discosrv-middleware"
+
+To send ``X-Client-Port``
```
Comment (tomasz1986) [inline users/stdiscosrv.rst:387]:
```suggestion
To send ``X-Client-Port``:
```

File: users/stdiscosrv.rst [hunk 4]
```diff
@@ -346,6 +346,99 @@ page. Note that that page is directed at setting up a proxy for the
...
+
+Traefik has ``X-Forwarded-Port`` which can only be changed using a
+plugin (added in version 2.5).
+
+1. Clone down a header rewriting plugin, the example here uses
+https://github.com/adyanth/header-transform
```
Comment (tomasz1986) [inline users/stdiscosrv.rst:393]:
```suggestion
https://github.com/adyanth/header-transform.

```

File: users/stdiscosrv.rst [hunk 5]
```diff
@@ -346,6 +346,99 @@ page. Note that that page is directed at setting up a proxy for the
...
+2. Mount the plugin into the Traefik Docker container using a volume and
+configure Traefik to read it
```
Comment (tomasz1986) [inline users/stdiscosrv.rst:395]:
```suggestion
configure Traefik to read it.
```

File: users/stdiscosrv.rst [hunk 6]
```diff
@@ -346,6 +346,99 @@ page. Note that that page is directed at setting up a proxy for the
...
+
+.. code-block:: yaml
+
+      traefik:
+        image: traefik:v2.5
+        # ...
+        command:
+          # ...
+          # Will read the plugin at /plugins-local/src/github.com/adyanth/header-transform and assign it
+          # to a middleware plugin named `header-transform-plugin`
```
Comment (tomasz1986) [inline users/stdiscosrv.rst:405]:
```suggestion
          # to a middleware plugin named `header-transform-plugin`.
```

File: users/stdiscosrv.rst [hunk 7]
```diff
@@ -346,6 +346,99 @@ page. Note that that page is directed at setting up a proxy for the
...
+          - "--experimental.localPlugins.header-transform-plugin.moduleName=github.com/adyanth/header-transform"
+          # ...
+        volumes:
+          # ...
+          # Traefik reads local plugins using a specific path, rooted at /plugins-local
```
Comment (tomasz1986) [inline users/stdiscosrv.rst:410]:
```suggestion
          # Traefik reads local plugins using a specific path, rooted at /plugins-local.
```

File: users/stdiscosrv.rst [hunk 8]
```diff
@@ -346,6 +346,99 @@ page. Note that that page is directed at setting up a proxy for the
...
+          # https://github.com/traefik/traefik/pull/8224
+          - "./traefik/header-transform:/plugins-local/src/github.com/adyanth/header-transform"
+          # ...
+
+3. In your dynamic config, define a middleware using the plugin which has
+a Rule that sets X-Client-Port to the value of X-Forwarded-Port
```
Comment (tomasz1986) [inline users/stdiscosrv.rst:416]:
```suggestion
a Rule that sets X-Client-Port to the value of X-Forwarded-Port.
```

File: users/stdiscosrv.rst [hunk 9]
```diff
@@ -346,6 +346,99 @@ page. Note that that page is directed at setting up a proxy for the
...
+            header-transform:
+              Rules:
+                - Rule:
+                  Name: 'X-Client-Port Set'
+                  Header: 'X-Client-Port'
+                  Value: '^X-Forwarded-Port'
+                  HeaderPrefix: "^"
+                  Type: 'Set'
+
+4. Add the middleware to the containers dynamic configuration
```
Comment (tomasz1986) [inline users/stdiscosrv.rst:433]:
```suggestion
4. Add the middleware to the containers dynamic configuration.
```

**Comment by tomasz1986** [comment]
No comments on the actual code, but I've suggested some punctuation fixes.

What's up with all the `# ...` comments though? Do they serve any practical purpose? If not, then I'd personally suggest to get rid of them, just to make the actual code shorter and cleaner.

**Comment by Cobertos** [comment]
Hmm, I added the `# ...` to omit things that aren't part of the header setup, but are still required to make a traefik/syncthing-discosrv container function. Just to show it's part of a larger whole.

I could update to make it a minimum viable config as-is? I don't think I can highlight/call out the specific lines that need to be added to make it work though then.

**Comment by tomasz1986** [comment]
> Hmm, I added the `# ...` to omit things that aren't part of the header setup, but are still required to make a traefik/syncthing-discosrv container function. Just to show it's part of a larger whole.

Ah, I see. It doesn't seem obvious on the first sight though, especially for someone new to the whole thing. If the larger whole is just too much content, how about at least a one-sentence summary of what is supposed to appear there?

File: users/stdiscosrv.rst [hunk 10]
```diff
@@ -346,6 +346,121 @@ page. Note that that page is directed at setting up a proxy for the
...
+"""""""""
+
+Traefik will send ``X-Forwarded-For`` by default.
+
+To send ``X-Forwarded-Tls-Client-Cert``:
+
+1. Add a `TLS Options <https://doc.traefik.io/traefik/https/tls/#client-authentication-mtls>`_
+that sets ``clientAuthType`` to ``RequireAnyClientCert`` in your dynamic
+config. Without this, Traefik will not pass the client certificate in the
+``X-Forwarded-Tls-Client-Cert`` header.
```
Comment (tomasz1986) [inline users/stdiscosrv.rst:359]:
```suggestion
#. Add a `TLS Options <https://doc.traefik.io/traefik/https/tls/#client-authentication-mtls>`_
   that sets ``clientAuthType`` to ``RequireAnyClientCert`` in your dynamic
   config. Without this, Traefik will not pass the client certificate in the
   ``X-Forwarded-Tls-Client-Cert`` header.
```

File: users/stdiscosrv.rst [hunk 11]
```diff
@@ -346,6 +346,121 @@ page. Note that that page is directed at setting up a proxy for the
...
+
+.. code-block:: yaml
+
+    tls:
+      options:
+        syncthing-discosrv:
+          clientAuth:
+            clientAuthType: RequireAnyClientCert
```
Comment (tomasz1986) [inline users/stdiscosrv.rst:367]:
```suggestion
    .. code-block:: yaml

       tls:
         options:
           syncthing-discosrv:
             clientAuth:
               clientAuthType: RequireAnyClientCert
```

File: users/stdiscosrv.rst [hunk 12]
```diff
@@ -346,6 +346,121 @@ page. Note that that page is directed at setting up a proxy for the
...
+
+2. Also in your dynamic config, add a middleware to pass the TLS client
+cert (`passtlsclientcert.pem=true`), add that to the router, and set your
+custom TLS options.
```
Comment (tomasz1986) [inline users/stdiscosrv.rst:371]:
```suggestion
#. Also in your dynamic config, add a middleware to pass the TLS client
   cert (`passtlsclientcert.pem=true`), add that to the router, and set your
   custom TLS options.
```

File: users/stdiscosrv.rst [hunk 13]
```diff
@@ -346,6 +346,121 @@ page. Note that that page is directed at setting up a proxy for the
...
+
+.. code-block:: yaml
```
Comment (tomasz1986) [inline users/stdiscosrv.rst:373]:
```suggestion
   .. code-block:: yaml
```

File: users/stdiscosrv.rst [hunk 14]
```diff
@@ -346,6 +346,121 @@ page. Note that that page is directed at setting up a proxy for the
...
+        command:
+          - "-http"
+
+To send ``X-Client-Port``:
+
+Traefik has ``X-Forwarded-Port`` which can only be changed using a
+plugin (added in version 2.5).
+
+1. Clone down a header rewriting plugin, the example here uses
+https://github.com/adyanth/header-transform.
```
Comment (tomasz1986) [inline users/stdiscosrv.rst:396]:
```suggestion
#. Clone down a header rewriting plugin, the example here uses
   https://github.com/adyanth/header-transform.
```

File: users/stdiscosrv.rst [hunk 15]
```diff
@@ -346,6 +346,121 @@ page. Note that that page is directed at setting up a proxy for the
...
+
+2. Mount the plugin into the Traefik Docker container using a volume and
+configure Traefik to read it.
```
Comment (tomasz1986) [inline users/stdiscosrv.rst:399]:
```suggestion
#. Mount the plugin into the Traefik Docker container using a volume and
   configure Traefik to read it.
```

File: users/stdiscosrv.rst [hunk 16]
```diff
@@ -346,6 +346,121 @@ page. Note that that page is directed at setting up a proxy for the
...
+
+.. code-block:: yaml
```
Comment (tomasz1986) [inline users/stdiscosrv.rst:401]:
```suggestion
   .. code-block:: yaml
```

File: users/stdiscosrv.rst [hunk 17]
```diff
@@ -346,6 +346,121 @@ page. Note that that page is directed at setting up a proxy for the
...
+          
+          # Traefik reads local plugins using a specific path, rooted at /plugins-local.
+          # https://github.com/traefik/traefik/pull/8224
+          - "./traefik/header-transform:/plugins-local/src/github.com/adyanth/header-transform"
+        ports:
+          - "80:80"
+          - "443:443"
+
+3. In your dynamic config, define a middleware using the plugin which has
+a Rule that sets X-Client-Port to the value of X-Forwarded-Port.
```
Comment (tomasz1986) [inline users/stdiscosrv.rst:427]:
```suggestion
#. In your dynamic config, define a middleware using the plugin which has
   a Rule that sets X-Client-Port to the value of X-Forwarded-Port.
```

File: users/stdiscosrv.rst [hunk 18]
```diff
@@ -346,6 +346,121 @@ page. Note that that page is directed at setting up a proxy for the
...
+        header-transform:
+          plugin:
+            header-transform:
+              Rules:
+                - Rule:
+                  Name: 'X-Client-Port Set'
+                  Header: 'X-Client-Port'
+                  Value: '^X-Forwarded-Port'
+                  HeaderPrefix: "^"
+                  Type: 'Set'
```
Comment (tomasz1986) [inline users/stdiscosrv.rst:442]:
```suggestion
   .. code-block:: yaml

      http:
        middlewares:
          header-transform:
            plugin:
              header-transform:
                Rules:
                  - Rule:
                    Name: 'X-Client-Port Set'
                    Header: 'X-Client-Port'
                    Value: '^X-Forwarded-Port'
                    HeaderPrefix: "^"
                    Type: 'Set'
```

File: users/stdiscosrv.rst [hunk 19]
```diff
@@ -346,6 +346,121 @@ page. Note that that page is directed at setting up a proxy for the
...
+
+4. Add the middleware to the containers dynamic configuration.
```
Comment (tomasz1986) [inline users/stdiscosrv.rst:444]:
```suggestion
#. Add the middleware to the containers dynamic configuration.
```

File: users/stdiscosrv.rst [hunk 20]
```diff
@@ -346,6 +346,121 @@ page. Note that that page is directed at setting up a proxy for the
...
+      - "traefik.http.middlewares.syncthing-discosrv-middleware.passtlsclientcert.pem=true"
+      - "traefik.http.services.syncthing-discosrv.loadbalancer.server.port=8443"
+      - "traefik.http.routers.syncthing-discosrv.entrypoints=https"
+      - "traefik.http.routers.syncthing-discosrv.rule=Host(`st-ds.xxx.dev`)"
+      - "traefik.http.routers.syncthing-discosrv.tls.options=syncthing-discosrv@file"
+      
+      # Add this line
+      - "traefik.http.routers.syncthing-discosrv.middlewares=syncthing-discosrv-middleware,header-transform@file"
+    command:
+      - "-http"
```
Comment (tomasz1986) [inline users/stdiscosrv.rst:463]:
```suggestion
   .. code-block:: yaml
 
     syncthing-discosrv:
       image: syncthing/discosrv
       volumes:
         - ./syncthing/DATA/syncthing-discosrv:/var/stdiscosrv
       labels:
         - "traefik.enable=true"
         - "traefik.http.middlewares.syncthing-discosrv-middleware.passtlsclientcert.pem=true"
         - "traefik.http.services.syncthing-discosrv.loadbalancer.server.port=8443"
         - "traefik.http.routers.syncthing-discosrv.entrypoints=https"
         - "traefik.http.routers.syncthing-discosrv.rule=Host(`st-ds.xxx.dev`)"
         - "traefik.http.routers.syncthing-discosrv.tls.options=syncthing-discosrv@file"
         
         # Add this line
         - "traefik.http.routers.syncthing-discosrv.middlewares=syncthing-discosrv-middleware,header-transform@file"
       command:
         - "-http"
```

**Comment by tomasz1986** [comment]
Looks good visually 👍, although I cannot comment on the code itself. You need to fix the spacing however, because the RST format uses `#.` for ordered lists, and the code between the items needs to be padded too.

You may want to have a look at https://sublime-and-sphinx-guide.readthedocs.io/en/latest/lists.html#code-images-and-other-content-inside-lists.

**Comment by Cobertos** [comment]
@kevdogg 
Yeah, that is correct, it shold be `header-transform-plugin` in middlewares section as you describe. I think I missed updating the name in the second location. Thanks, I'll fix

**Comment by Cobertos** [comment]
Kk, fixed the name to `header-transform-plugin`. Also hopefully fixed the spacing? I used 3 spaces for all the code blocks inside ordered lists, which is what the docs wanted (though the actual code from the docs seems to use 4 spaces?)

File: users/stdiscosrv.rst [hunk 21]
```diff
@@ -346,6 +346,121 @@ page. Note that that page is directed at setting up a proxy for the
...
+   config. Without this, Traefik will not pass the client certificate in the
+   ``X-Forwarded-Tls-Client-Cert`` header.
+
+   .. code-block:: yaml
+
+     tls:
+       options:
+         syncthing-discosrv:
+           clientAuth:
+             clientAuthType: RequireAnyClientCert
```
Comment (AudriusButkevicius) [inline users/stdiscosrv.rst:367]: In theory, this is only needed for posts, and not gets.

File: users/stdiscosrv.rst [hunk 22]
```diff
@@ -346,6 +346,121 @@ page. Note that that page is directed at setting up a proxy for the
...
+         - "traefik.http.routers.syncthing-discosrv.rule=Host(`st-ds.xxx.dev`)"
+         - "traefik.http.routers.syncthing-discosrv.tls.options=syncthing-discosrv@file"
+         - "traefik.http.routers.syncthing-discosrv.middlewares=syncthing-discosrv-middleware"
+       command:
+         - "-http"
+
+To send ``X-Client-Port``:
+
+Traefik has ``X-Forwarded-Port`` which can only be changed using a
+plugin (plugins were added in version 2.5 of Traefik).
```
Comment (calmh) [inline users/stdiscosrv.rst:393]: For what it's worth I think this sections is wrong. `X-Forwarded-Port` is the port number on the proxy that received the request, i.e., most likely `443`. The `X-Client-Port` we want is the source port on the client side. If that's not available that's mostly fine, and this section can be skipped.

**Comment by st-review** [comment]
:robot: *beep boop*

I'm going to close this pull request as it has been idle for more than 90 days.

This is not a rejection, merely a removal from the list of active pull requests that is periodically reviewed by humans. The pull request can be reopened when there is new activity, and merged once any remaining issues are resolved.

**Comment by bugsyb** [comment]
@calmh , @tomasz1986 @kevdogg - could this PR be merged please? It took a while to land here and Traefik info from @Cobertos  is really helpful.

**Comment by AudriusButkevicius** [comment]
The comments need addressing before this can be merged.

**Pull request closed without merging**